File: waterBottles2.go
func maxBottlesDrunk(numBottles int, numExchange int) int {
    drank, full, empty := 0, numBottles, 0
    for full > 0 {
        drank += full
        empty += full
        full = 0
        if empty >= numExchange {
            empty -= numExchange
            full += 1
            numExchange++
        }
    }
    return drank
}

File: waterBottles2.py
class Solution:
    def maxBottlesDrunk(self, numBottles: int, numExchange: int) -> int:
        bottles_drunk = 0
        full_bottles = numBottles
        empty_bottles = 0
        
        while full_bottles > 0:
            # Drink all full bottles
            bottles_drunk += full_bottles
            empty_bottles += full_bottles
            
            # Convert empty bottles to full bottles
            full_bottles = empty_bottles // numExchange
            empty_bottles = empty_bottles % numExchange
            
            # Increase numExchange by 1 after each exchange
            numExchange += 1
            
        return bottles_drunk
